cpp/src/c_api/legacy_fa2.cpp

```diff
@@ -0,0 +1,291 @@
+
+      rmm::device_uvector<weight_t> tmp_weights(0, handle_.get_stream());
+      if (edge_weights == nullptr) {
+        tmp_weights.resize(edge_partition_view.indices().size(), handle_.get_stream());
+        cugraph::detail::scalar_fill(handle_, tmp_weights.data(), tmp_weights.size(), weight_t{1});
+      }
+
+      // Decompress to edgelist
+
+      auto [srcs, dsts, wgts, edge_ids, edge_types] =
+        cugraph::decompress_to_edgelist<vertex_t, edge_t, weight_t, int32_t, false, multi_gpu>(
+          handle_,
+          graph_view,
+          (edge_weights != nullptr) ? std::make_optional(edge_weights->view()) : std::nullopt,
+          std::nullopt,
+          std::nullopt,
+          (number_map != nullptr) ? std::make_optional(raft::device_span<vertex_t const>{
+                                      number_map->data(), number_map->size()})
+                                  : std::nullopt);
+
+      cugraph::legacy::GraphCOOView<vertex_t, edge_t, weight_t> legacy_coo_graph_view(
+        const_cast<vertex_t*>(srcs.data()),
+        const_cast<vertex_t*>(dsts.data()),
+        (edge_weights == nullptr) ? tmp_weights.data() : const_cast<weight_t*>(wgts->data()),
+        graph->number_of_vertices(),
+        edge_partition_view.number_of_edges());
+
+      cugraph::internals::GraphBasedDimRedCallback* callback = nullptr;
+
+      rmm::device_uvector<vertex_t> vertices(graph_view.local_vertex_partition_range_size(),
+                                             handle_.get_stream());
+      raft::copy(vertices.data(), number_map->data(), vertices.size(), handle_.get_stream());
+
+      rmm::device_uvector<float> pos(2 * (edge_partition_view.offsets().size() - 1),
+                                     handle_.get_stream());
+
+      cugraph::force_atlas2<vertex_t, edge_t, weight_t>(
+        handle_,
+        // rng_state_->rng_state_, # FIXME: Add support
+        legacy_coo_graph_view,
+        pos.data(),
+        max_iter_,
+        x_start_ != nullptr ? x_start_->as_type<float>() : nullptr,
+        y_start_ != nullptr ? y_start_->as_type<float>() : nullptr,
+        outbound_attraction_distribution_,
+        lin_log_mode_,
+        prevent_overlapping_,
+        edge_weight_influence_,
+        jitter_tolerance_,
+        barnes_hut_optimize_,
+        barnes_hut_theta_,
+        scaling_ratio_,
+        strong_gravity_mode_,
+        gravity_,
+        verbose_,
+        callback);
+
+      // result_ = new cugraph::c_api::cugraph_type_erased_device_array_t(pos,
+      // graph_->weight_type_);
+
+      rmm::device_uvector<float> x_axis(graph_view.local_vertex_partition_range_size(),
+                                        handle_.get_stream());
+
+      rmm::device_uvector<float> y_axis(graph_view.local_vertex_partition_range_size(),
+                                        handle_.get_stream());
+
+      raft::copy(x_axis.data(), pos.data(), x_axis.size(), handle_.get_stream());
```

Comment:
Is this correct?  I thought the x and y values are interleaved.  You would need a strided copy (elements 0, 2, 4, ... go in x_axis 0,1,2... and elements 1, 3, 5, ... go in y_axis 0, 1, 2...).  You might double check that.

Reply:
I double checked and I don't think they are interleaved. The array pos holds both the x and y coordinate and when called for instance in [apply_forces](https://github.com/rapidsai/cugraph/blob/branch-25.04/cpp/src/layout/legacy/exact_fa2.cuh#L173:#L174) (pos, pos+n) where n is the number of vertex, this translates to the beginning of the [x and y position](https://github.com/rapidsai/cugraph/blob/branch-25.04/cpp/src/layout/legacy/fa2_kernels.cuh#L162:L163)

I also made sure the result matches the non PLC version which was the case